src/compiler/utilities.ts

```diff
@@ -3467,27 +3467,107 @@ namespace ts {
         return 0;
     }
 
-    export function levenshtein(s1: string, s2: string): number {
-        let previous: number[] = new Array(s2.length + 1);
-        let current: number[] = new Array(s2.length + 1);
-        for (let i = 0; i < s2.length + 1; i++) {
+    /**
+     * Given a name and a list of symbols whose names are *not* equal to the name, return a spelling suggestion if there is one that is close enough.
+     * Names less than length 3 only check for case-insensitive equality, not levenshtein distance.
+     *
+     * If there is a candidate that's the same except for case, return that.
+     * If there is a candidate that's within one edit of the name, return that.
+     * Otherwise, return the candidate with the smallest Levenshtein distance,
+     *    except for candidates:
+     *      * With no name
+     *      * Whose meaning doesn't match the `meaning` parameter.
+     *      * Whose length differs from the target name by more than 0.34 of the length of the name.
+     *      * Whose levenshtein distance is more than 0.4 of the length of the name
+     *        (0.4 allows 1 substitution/transposition for every 5 characters,
+     *         and 1 insertion/deletion at 3 characters)
+     */
+    export function getSpellingSuggestionForName(name: string, symbols: Symbol[], meaning: SymbolFlags): Symbol | undefined {
+        const maximumLengthDifference = Math.min(2, Math.floor(name.length * 0.34));
+        let bestDistance = Math.floor(name.length * 0.4);
+        let bestCandidate: Symbol | undefined;
+        let justCheckExactMatches = false;
+        const nameLowerCase = name.toLowerCase();
+        for (const candidate of symbols) {
+            const candidateName = symbolName(candidate);
+            if (!(candidate.flags & meaning && Math.abs(candidateName.length - nameLowerCase.length) <= maximumLengthDifference)) {
+                continue;
+            }
+            const candidateNameLowerCase = candidateName.toLowerCase();
+            if (candidateNameLowerCase === nameLowerCase) {
+                return candidate;
+            }
+            if (justCheckExactMatches) {
+                continue;
+            }
+            if (candidateNameLowerCase.length < 3 ||
+                nameLowerCase.length < 3 ||
+                candidateNameLowerCase === "eval" ||
+                candidateNameLowerCase === "intl" ||
+                candidateNameLowerCase === "undefined" ||

```

Comment:
Why are these excluded? I'd really like to be able to automatically fix `undefnied`.

Reply:
The suggestions were not good at one point. That might have changed during final tweaks to the numbers, because I ended up less likely to give suggestions for short identifiers than when I was mid-development. That's specifically why the 3-4 character identifiers are here. Maybe `undefined` is no longer affected.

Coverage is pretty good at this point, so if you remove items from this list, you should see what bad suggestions happen.

Reply:
(If any; it may be fine to suggest `undefined` now)